### PR TITLE
[util] Add script and first template to autogen (some) testutils.

### PR DESCRIPTION
**Pull request by timothytrippel**
This partially addresses #9038 by auto-generating a higher-level
testutils library to service ISRs. Specifically this commit:

1. adds a Python script, `util/autogen_testutils.py`, that renders any
   testutils templates that live in util/autogen_testutils/, and
2. adds a `isr_testutils.h.tpl` header template.

Signed-off-by: Timothy Trippel <ttrippel@google.com>

### CODE REVIEW

**Comment by sriyerg** [comment]
Ohh, we are actually doing this? It was just an idea I wanted to float.

**Comment by timothytrippel** [comment]
> Ohh, we are actually doing this? It was just an idea I wanted to float.

I can mark as draft PR and we can discuss. Wanted to get a PR out to demonstrate _how_ it would be done.

File: util/autogen_testutils.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                testutils_template.render(
+                    ips_with_difs=ips_with_difs,
+                    autogen_banner=get_autogen_banner(
+                        "util/autogen_testutils.py", comment="//"),
+                    license_banner=get_license_banner(comment="//")))
+
+            # Format autogenerated file with clang-format.
+            assert (shutil.which("clang-format") and
+                    "ERROR: clang-format must be installed to format "
+                    " autogenerated code to pass OpenTitan CI checks.")
```
Comment (matutem) [inline util/autogen_testutils.py:75]: This could be done prior to the loop and fail early.

File: util/autogen_testutils/isr_testutils.h.tpl [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+   * Services an ${ip.name_snake} IRQ at the IP that raised it, and verifies this
+   * matches the IRQ that was raised at the PLIC.
+   *  
+   * @param ${ip.name_snake} A(n) ${ip.name_snake} DIF handle. 
+   * @param plic_irq_id The triggered PLIC IRQ ID. 
+   * @param plic_${ip.name_snake}_start_irq_id The PLIC IRQ ID where ${ip.name_snake} starts. 
+   * @param expected_${ip.name_snake}_irq The expected ${ip.name_snake} IRQ. 
+   * @param is_only_irq This is the only IRQ expected to be raised.
+   */
+  void isr_testutils_${ip.name_snake}(dif_${ip.name_snake}_t *${ip.name_snake},
```
Comment (matutem) [inline util/autogen_testutils/isr_testutils.h.tpl:20]: What about ${ip.name_snake}_isr_testutils?

**Review by matutem — COMMENTED** [review]
Couple comments

File: util/autogen_testutils/isr_testutils.h.tpl (hunk 2 above)
Comment (timothytrippel) [reply util/autogen_testutils/isr_testutils.h.tpl:20]: The way it is keeps the convention of: <IP or functionality lib name>_testutils_<function name>, described in #7976, where "<IP or functionality lib name>_testutils" is supposed to match the testutils file name.

File: util/autogen_testutils.py (hunk 1 above)
Comment (timothytrippel) [reply util/autogen_testutils.py]: Fixed.

**Review by mcy — APPROVED** [review]

File: util/autogen_testutils.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import subprocess
+import sys
+from pathlib import Path
+
+from mako.template import Template
+
+from autogen_banner import get_autogen_banner, get_license_banner
+from make_new_dif.ip import Ip
+
+# This file is $REPO_TOP/util/make_new_dif.py, so it takes two parent()
```
[the remark this answers is not in the document]
Comment (timothytrippel) [reply util/autogen_testutils.py]: Nice catch! Fixed.

**Pull request merged**